### PR TITLE
2X training speed improvement: prefetch next batch from the data loader while running current batch

### PR DESCRIPTION
**Pull request by koush**
This improves my training speed by 2x. This fetches the next batch of images while processing the current batch. Results may vary based on system.

My setup:
2x NVIDIA 4090 
10900k
64GB RAM
M.2 Storage for coco dataset

Epoch time: 34 minutes to 14 minutes.

### CODE REVIEW

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 1]
```diff
@@ -483,7 +483,31 @@ def _train_epoch(self, context: PhaseContext, silent_mode: bool = False) -> tupl
...
+
+                def __next__(self):
+                    value = self.prefetch_future.result()
+                    self.prefetch()
+                    return value
+
+                def close(self):
+                    self.executor.shutdown()
+
+            for batch_idx, batch_items in PrefetchIterator(enumerate(progress_bar_train_loader)):
```
Comment (BloodAxe) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:510]: Shouldn't this be reverved? E.g `enumerate(PrefetchIterator(progress_bar_train_loader))`
I imagine we want to update tqdm only from main process once we are processing batches and not during prefetch.

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 2]
```diff
@@ -483,7 +483,31 @@ def _train_epoch(self, context: PhaseContext, silent_mode: bool = False) -> tupl
 
             context.update_context(loss_avg_meter=loss_avg_meter, metrics_compute_fn=self.train_metrics)
 
-            for batch_idx, batch_items in enumerate(progress_bar_train_loader):
+            class PrefetchIterator:
```
Comment (BloodAxe) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:486]: Very nice trick!
Could you please move this class somewhere outside the Trainer and use it in both train & validation loops?

File: src/super_gradients/training/sg_trainer/sg_trainer.py [hunk 3]
```diff
@@ -483,7 +483,31 @@ def _train_epoch(self, context: PhaseContext, silent_mode: bool = False) -> tupl
...
+
+                def __iter__(self):
+                    return self
+
+                def __next__(self):
+                    value = self.prefetch_future.result()
+                    self.prefetch()
+                    return value
+
+                def close(self):
```
Comment (BloodAxe) [inline src/super_gradients/training/sg_trainer/sg_trainer.py:507]: Who is calling `close` method? Maybe we should have some __enter__ / __exit__ methods to ensure we do proper cleanup of the ThreadPoolExecutor.

Comment (koush) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:507]: Close is automatically called when the for loop using the iterator breaks.

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 1 above)
Comment (koush) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:510]: I believe you mean the prefetch iterator should wrap the data loader, and not the progress bar loader, yes.

**Comment by BloodAxe** [comment]
It seems your build fails due to code formatting issues. Please check https://github.com/Deci-AI/super-gradients/blob/master/CONTRIBUTING.md to learn how to configure pre-commit hooks and signed commits.

File: src/super_gradients/training/sg_trainer/sg_trainer.py (hunk 3 above)
Comment (koush) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:507]: Actually I was mistaken, the python language construct to release an iterator is `__del__`, not `close`.

Comment (koush) [reply src/super_gradients/training/sg_trainer/sg_trainer.py:507]: Changed how the iterator is created/managed so this code is defunct.

**Comment by koush** [comment]
> It seems your build fails due to code formatting issues. Please check https://github.com/Deci-AI/super-gradients/blob/master/CONTRIBUTING.md to learn how to configure pre-commit hooks and signed commits.

fixed with the precommit hook and black

**Comment by koush** [comment]
I think this is failing due to my force push. I'll recreate the pull request.
